File: Sources/PomodoroTimer.swift
import Foundation
import AppKit
import UserNotifications

enum CycleType: String {
    case work       = "TRABAJO"
    case shortBreak = "DESCANSO CORTO"
    case longBreak  = "DESCANSO LARGO"
}

@MainActor
final class PomodoroTimer: ObservableObject {

    // MARK: - Published state

    @Published var timeRemaining: TimeInterval
    @Published var currentCycle: CycleType = .work
    @Published var isRunning = false
    @Published var completedSessions = 0
    @Published var progress: Double = 1.0
    @Published var isOvertime = false
    @Published var overtimeSeconds: TimeInterval = 0

    // MARK: - Settings (auto-persisted via didSet)

    @Published var workDuration: TimeInterval {
        didSet { UserDefaults.standard.set(workDuration, forKey: "workDuration") }
    }
    @Published var shortBreakDuration: TimeInterval {
        didSet { UserDefaults.standard.set(shortBreakDuration, forKey: "shortBreakDuration") }
    }
    @Published var longBreakDuration: TimeInterval {
        didSet { UserDefaults.standard.set(longBreakDuration, forKey: "longBreakDuration") }
    }
    @Published var sessionsUntilLongBreak: Int {
        didSet { UserDefaults.standard.set(sessionsUntilLongBreak, forKey: "sessionsUntilLongBreak") }
    }
    @Published var autoStart: Bool {
        didSet { UserDefaults.standard.set(autoStart, forKey: "autoStart") }
    }

    // MARK: - Overtime stats (auto-persisted via didSet)

    @Published var totalWorkOvertime: TimeInterval {
        didSet { UserDefaults.standard.set(totalWorkOvertime, forKey: "totalWorkOvertime") }
    }
    @Published var totalBreakOvertime: TimeInterval {
        didSet { UserDefaults.standard.set(totalBreakOvertime, forKey: "totalBreakOvertime") }
    }

    // MARK: - Private

    private var timerSource: Timer?
    private var cycleEndDate: Date?   // absolute moment the cycle should end
    private var pauseDate: Date?      // moment we paused (nil when running)
    private var totalDuration: TimeInterval

    // MARK: - Init

    init() {
        let d = UserDefaults.standard
        let work     = d.object(forKey: "workDuration")           as? TimeInterval ?? 25 * 60
        let short    = d.object(forKey: "shortBreakDuration")     as? TimeInterval ?? 5 * 60
        let long     = d.object(forKey: "longBreakDuration")      as? TimeInterval ?? 15 * 60
        let sessions = d.object(forKey: "sessionsUntilLongBreak") as? Int          ?? 4
        let auto     = d.object(forKey: "autoStart")              as? Bool         ?? false
        let wOT      = d.object(forKey: "totalWorkOvertime")      as? TimeInterval ?? 0
        let bOT      = d.object(forKey: "totalBreakOvertime")     as? TimeInterval ?? 0

        self.workDuration           = work
        self.shortBreakDuration     = short
        self.longBreakDuration      = long
        self.sessionsUntilLongBreak = sessions
        self.autoStart              = auto
        self.totalWorkOvertime      = wOT
        self.totalBreakOvertime     = bOT
        self.timeRemaining          = work
        self.totalDuration          = work
    }

    // MARK: - Computed

    var currentCycleDuration: TimeInterval {
        switch currentCycle {
        case .work:       return workDuration
        case .shortBreak: return shortBreakDuration
        case .longBreak:  return longBreakDuration
        }
    }

    var formattedTime: String {
        if isOvertime {
            let secs = Int(overtimeSeconds)
            return String(format: "+%02d:%02d", secs / 60, secs % 60)
        }
        let total = Int(ceil(timeRemaining))
        return String(format: "%02d:%02d", total / 60, total % 60)
    }

    // Preview the next cycle without mutating state (used for notification on overtime entry)
    private var peekNextCycle: CycleType {
        switch currentCycle {
        case .work:
            let next = completedSessions + 1
            guard sessionsUntilLongBreak > 0 else { return .shortBreak }
            return (next % sessionsUntilLongBreak == 0) ? .longBreak : .shortBreak
        case .shortBreak, .longBreak:
            return .work
        }
    }

    // MARK: - Public controls

    func togglePlayPause() { isRunning ? pause() : start() }

    func start() {
        guard !isRunning else { return }

        if let pd = pauseDate, let endDate = cycleEndDate {
            // Resume: shift the end date forward by the time we were paused
            // This works correctly both in normal mode and in overtime mode.
            let pausedFor = Date().timeIntervalSince(pd)
            cycleEndDate = endDate.addingTimeInterval(pausedFor)
            pauseDate = nil
        } else {
            // Fresh start
            pauseDate = nil
            cycleEndDate = Date().addingTimeInterval(timeRemaining)
        }

        isRunning = true
        timerSource = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in self?.tick() }
        }
        RunLoop.main.add(timerSource!, forMode: .common)
    }

    func pause() {
        guard isRunning else { return }
        isRunning = false
        pauseDate = Date()
        timerSource?.invalidate()
        timerSource = nil
    }

    func reset() {
        isRunning = false
        timerSource?.invalidate()
        timerSource = nil
        cycleEndDate = nil
        pauseDate = nil
        isOvertime = false
        overtimeSeconds = 0
        timeRemaining = currentCycleDuration
        totalDuration = currentCycleDuration
        progress = 1.0
    }

    func skip() {
        // Count session only if the timer reached 0 naturally (overtime was active)
        let shouldCount = isOvertime && currentCycle == .work
        accumulateOvertime()

        isRunning = false
        timerSource?.invalidate()
        timerSource = nil
        cycleEndDate = nil
        pauseDate = nil
        isOvertime = false
        overtimeSeconds = 0

        advance(countSession: shouldCount)
        if autoStart { start() }
    }

    func applySettings() {
        if !isRunning {
            cycleEndDate = nil
            pauseDate = nil
            isOvertime = false
            overtimeSeconds = 0
            totalDuration = currentCycleDuration
            timeRemaining = currentCycleDuration
            progress = 1.0
        }
    }

    // MARK: - Private

    private func tick() {
        guard let endDate = cycleEndDate else { return }

        // diff > 0: time still remaining. diff < 0: we're in overtime.
        let diff = endDate.timeIntervalSinceNow

        if diff > 0 {
            timeRemaining = diff
            progress = totalDuration > 0 ? diff / totalDuration : 0
        } else {
            timeRemaining = 0
            progress = 0

            if !isOvertime {
                // First tick past 0 — enter overtime
                isOvertime = true
                playAlarm()
                scheduleNotification(for: peekNextCycle)
            }
            overtimeSeconds = -diff   // diff is negative, so -diff is positive
        }
    }

    private func advance(countSession: Bool) {
        switch currentCycle {
        case .work:
            if countSession { completedSessions += 1 }
            let goLong = completedSessions > 0 && completedSessions % sessionsUntilLongBreak == 0
            currentCycle = goLong ? .longBreak : .shortBreak
        case .shortBreak, .longBreak:
            currentCycle = .work
        }
        totalDuration = currentCycleDuration
        timeRemaining = currentCycleDuration
        progress = 1.0
    }

    // Add current overtime to the running total before clearing it.
    private func accumulateOvertime() {
        guard isOvertime && overtimeSeconds > 0 else { return }
        if currentCycle == .work {
            totalWorkOvertime += overtimeSeconds
        } else {
            totalBreakOvertime += overtimeSeconds
        }
    }

    private func playAlarm() {
        (NSSound(named: NSSound.Name("Glass")) ?? NSSound(named: NSSound.Name("Ping")))?.play()
    }

    private func scheduleNotification(for next: CycleType) {
        let content   = UNMutableNotificationContent()
        content.title = next == .work ? "¡A trabajar!" : "¡Es hora de descansar!"
        content.body  = next.rawValue
        content.sound = nil

        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 0.5, repeats: false)
        let request = UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: trigger)
        UNUserNotificationCenter.current().add(request)
    }
}

File: flutter/lib/models/pomodoro_timer.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

enum CycleType {
  work,
  shortBreak,
  longBreak;

  String get label {
    switch (this) {
      case CycleType.work:       return 'TRABAJO';
      case CycleType.shortBreak: return 'DESCANSO CORTO';
      case CycleType.longBreak:  return 'DESCANSO LARGO';
    }
  }
}

class PomodoroTimer extends ChangeNotifier {

  // ── Settings ─────────────────────────────────────────────────
  int workMinutes;
  int shortBreakMinutes;
  int longBreakMinutes;
  int sessionsUntilLongBreak;
  bool autoStart;

  // ── Overtime stats ───────────────────────────────────────────
  Duration totalWorkOvertime;
  Duration totalBreakOvertime;

  // ── Private state ────────────────────────────────────────────
  CycleType _currentCycle = CycleType.work;
  bool _isRunning = false;
  int _completedSessions = 0;
  Duration _timeRemaining;
  Duration _totalDuration;
  bool _isOvertime = false;
  Duration _overtimeElapsed = Duration.zero;

  Timer? _ticker;
  DateTime? _cycleEndDate;  // absolute moment the cycle should end
  DateTime? _pauseDate;     // moment we paused (null when running)

  void Function(CycleType next)? onCycleComplete;

  PomodoroTimer({
    this.workMinutes = 25,
    this.shortBreakMinutes = 5,
    this.longBreakMinutes = 15,
    this.sessionsUntilLongBreak = 4,
    this.autoStart = false,
    this.totalWorkOvertime = Duration.zero,
    this.totalBreakOvertime = Duration.zero,
  })  : _timeRemaining = const Duration(minutes: 25),
        _totalDuration = const Duration(minutes: 25);

  // ── Getters ──────────────────────────────────────────────────

  CycleType get currentCycle      => _currentCycle;
  bool      get isRunning         => _isRunning;
  int       get completedSessions => _completedSessions;
  bool      get isOvertime        => _isOvertime;
  Duration  get overtimeElapsed   => _overtimeElapsed;

  double get progress {
    if (_isOvertime) return 0.0;
    final total = _totalDuration.inMilliseconds;
    return total == 0 ? 1.0 : (_timeRemaining.inMilliseconds / total).clamp(0.0, 1.0);
  }

  String get formattedTime {
    if (_isOvertime) {
      final s = _overtimeElapsed.inSeconds;
      return '+${(s ~/ 60).toString().padLeft(2, '0')}:${(s % 60).toString().padLeft(2, '0')}';
    }
    final m = _timeRemaining.inMinutes.remainder(60).toString().padLeft(2, '0');
    final s = _timeRemaining.inSeconds.remainder(60).toString().padLeft(2, '0');
    return '$m:$s';
  }

  int get filledDots {
    if (sessionsUntilLongBreak == 0) return 0;
    return _completedSessions % sessionsUntilLongBreak;
  }

  Duration get _cycleDuration {
    switch (_currentCycle) {
      case CycleType.work:       return Duration(minutes: workMinutes);
      case CycleType.shortBreak: return Duration(minutes: shortBreakMinutes);
      case CycleType.longBreak:  return Duration(minutes: longBreakMinutes);
    }
  }

  // Preview the next cycle without mutating state (used for notification on overtime entry)
  CycleType get _peekNextCycle {
    switch (_currentCycle) {
      case CycleType.work:
        if (sessionsUntilLongBreak == 0) return CycleType.shortBreak;
        final next = _completedSessions + 1;
        return (next % sessionsUntilLongBreak == 0) ? CycleType.longBreak : CycleType.shortBreak;
      case CycleType.shortBreak:
      case CycleType.longBreak:
        return CycleType.work;
    }
  }

  // ── Factory ──────────────────────────────────────────────────

  static Future<PomodoroTimer> load() async {
    final p = await SharedPreferences.getInstance();
    return PomodoroTimer(
      workMinutes:            p.getInt('workMinutes')              ?? 25,
      shortBreakMinutes:      p.getInt('shortBreakMinutes')        ?? 5,
      longBreakMinutes:       p.getInt('longBreakMinutes')         ?? 15,
      sessionsUntilLongBreak: p.getInt('sessionsUntilLongBreak')   ?? 4,
      autoStart:              p.getBool('autoStart')               ?? false,
      totalWorkOvertime:      Duration(seconds: p.getInt('totalWorkOvertimeSecs')  ?? 0),
      totalBreakOvertime:     Duration(seconds: p.getInt('totalBreakOvertimeSecs') ?? 0),
    );
  }

  Future<void> saveSettings() async {
    final p = await SharedPreferences.getInstance();
    await p.setInt('workMinutes', workMinutes);
    await p.setInt('shortBreakMinutes', shortBreakMinutes);
    await p.setInt('longBreakMinutes', longBreakMinutes);
    await p.setInt('sessionsUntilLongBreak', sessionsUntilLongBreak);
    await p.setBool('autoStart', autoStart);
    await p.setInt('totalWorkOvertimeSecs',  totalWorkOvertime.inSeconds);
    await p.setInt('totalBreakOvertimeSecs', totalBreakOvertime.inSeconds);
  }

  // ── Controls ─────────────────────────────────────────────────

  void togglePlayPause() => _isRunning ? pause() : start();

  void start() {
    if (_isRunning) return;

    final now = DateTime.now();
    if (_pauseDate != null && _cycleEndDate != null) {
      // Resume: shift the end date forward by the time we were paused.
      // Works correctly both in normal mode and in overtime (cycleEndDate stays in the past,
      // and the shift keeps the overtime duration where we left off).
      final pausedFor = now.difference(_pauseDate!);
      _cycleEndDate = _cycleEndDate!.add(pausedFor);
      _pauseDate = null;
    } else {
      // Fresh start
      _pauseDate = null;
      _cycleEndDate = now.add(_timeRemaining);
    }

    _isRunning = true;
    _ticker = Timer.periodic(const Duration(milliseconds: 500), (_) => _tick());
    notifyListeners();
  }

  void pause() {
    if (!_isRunning) return;
    _isRunning = false;
    _pauseDate = DateTime.now();
    _ticker?.cancel();
    _ticker = null;
    notifyListeners();
  }

  void reset() {
    _isRunning = false;
    _ticker?.cancel();
    _ticker = null;
    _cycleEndDate = null;
    _pauseDate = null;
    _isOvertime = false;
    _overtimeElapsed = Duration.zero;
    _timeRemaining = _cycleDuration;
    _totalDuration = _cycleDuration;
    notifyListeners();
  }

  void skip() {
    // Count the session only if the timer reached 0 naturally (overtime was active)
    final shouldCount = _isOvertime && _currentCycle == CycleType.work;
    _accumulateOvertime();

    _isRunning = false;
    _ticker?.cancel();
    _ticker = null;
    _cycleEndDate = null;
    _pauseDate = null;
    _isOvertime = false;
    _overtimeElapsed = Duration.zero;

    _advance(countSession: shouldCount);
    if (autoStart) start();
  }

  void fullReset() {
    _isRunning = false;
    _ticker?.cancel();
    _ticker = null;
    _cycleEndDate = null;
    _pauseDate = null;
    _isOvertime = false;
    _overtimeElapsed = Duration.zero;

    _currentCycle = CycleType.work;
    _timeRemaining = Duration(minutes: workMinutes);
    _totalDuration = Duration(minutes: workMinutes);

    _completedSessions = 0;
    totalWorkOvertime = Duration.zero;
    totalBreakOvertime = Duration.zero;

    saveSettings(); // fire-and-forget, igual que en _accumulateOvertime()
    notifyListeners();
  }

  void applySettings() {
    saveSettings();
    if (!_isRunning) {
      _cycleEndDate = null;
      _pauseDate = null;
      _isOvertime = false;
      _overtimeElapsed = Duration.zero;
      _timeRemaining = _cycleDuration;
      _totalDuration = _cycleDuration;
      notifyListeners();
    }
  }

  // ── Private ──────────────────────────────────────────────────

  void _tick() {
    if (_cycleEndDate == null) return;

    // diff > 0: time still remaining. diff < 0: we're in overtime.
    final diff = _cycleEndDate!.difference(DateTime.now());

    if (diff > Duration.zero) {
      _timeRemaining = diff;
      notifyListeners();
    } else {
      _timeRemaining = Duration.zero;
      final elapsed = Duration(microseconds: diff.inMicroseconds.abs());

      if (!_isOvertime) {
        // First tick past 0 — enter overtime
        _isOvertime = true;
        onCycleComplete?.call(_peekNextCycle);
      }
      _overtimeElapsed = elapsed;
      notifyListeners();
    }
  }

  void _advance({required bool countSession}) {
    switch (_currentCycle) {
      case CycleType.work:
        if (countSession) _completedSessions++;
        final goLong = _completedSessions > 0 &&
            _completedSessions % sessionsUntilLongBreak == 0;
        _currentCycle = goLong ? CycleType.longBreak : CycleType.shortBreak;
      case CycleType.shortBreak:
      case CycleType.longBreak:
        _currentCycle = CycleType.work;
    }
    _totalDuration = _cycleDuration;
    _timeRemaining = _cycleDuration;
    notifyListeners();
  }

  // Add current overtime to the running total before clearing it.
  void _accumulateOvertime() {
    if (!_isOvertime || _overtimeElapsed == Duration.zero) return;
    if (_currentCycle == CycleType.work) {
      totalWorkOvertime += _overtimeElapsed;
    } else {
      totalBreakOvertime += _overtimeElapsed;
    }
    saveSettings(); // fire-and-forget; persists updated stats
  }

  @override
  void dispose() {
    _ticker?.cancel();
    super.dispose();
  }
}
